### PR TITLE
[Transform] Fix transform fails when getting field mappings

### LINKED ISSUE #75693: [Transform] Transform fails with "Failed to gather field mappings"
**Affected version**: 7.5-
**Fixed with**: tbd

Transform can fail on start or at runtime in case the transform process gets relocated to another node. The reason for the relocation can be a rolling upgrade, a rolling restart or a temporary instability in the cluster.

This also affects a transform included in endpoint fleet package: https://github.com/elastic/kibana/issues/104624

**Mitigation**

Force stop the transform and start it again, either in the UI or using:

```
POST _transform/{FAILING_TRANSFORM}/_stop?force=true
POST _transform/{FAILING_TRANSFORM}/_start
```

**Solution**

Transform requires the field mappings of the destination index for writing output documents. The failure happens for the call to gather these mappings. In some situations, especially on rolling upgrades, the destination index isn't ready at the time of this call. This leads to a failure and the transform is set to `FAILED`.

The root problem is the place where we get these mappings, which is at the start of the task. This is unfortunately a point of no return. It's not possible to retry or wait for the destination index in the start phase.

Getting the field mappings should be moved into the transform indexer, this will allow the indexer to retry the operation. In addition this fixes another issue: if the destination index is changed using `_update` the field mappings aren't re-retrieved.

### PR DESCRIPTION
**Pull request by hendrikmuhs**
move field mapping retrieval from task startup into indexer

fixes #75693

### CODE REVIEW

**Comment by hendrikmuhs** [comment]
FWIW: I think we should move more parts from the startup phase of the task into the indexer. But as follow up PR.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 1]
```diff
@@ -100,7 +100,7 @@
 
     protected final AtomicReference<Collection<ActionListener<Void>>> saveStateListeners = new AtomicReference<>();
 
-    private final Map<String, String> fieldMappings;
+    private Map<String, String> fieldMappings;
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:103]: does this need to be `volatile`?

**Review by benwtrent — APPROVED** [review]
I think the change is good. One question around memory volatility.

**Review by przemekwitek — APPROVED** [review]
LGTM

**Pull request merged**